contracts/InterchainTokenService.sol

```diff
@@ -1019,4 +982,37 @@ contract InterchainTokenService is
 
         _callContract(destinationChain, payload, msg.value);
     }
+
+    function _takeToken(bytes32 tokenId, address from, uint256 amount) internal returns (uint256, address tokenAddress) {
+        address tokenManager_ = tokenManagerAddress(tokenId);
+        uint256 tmt;
+        (tmt, tokenAddress) = ITokenManagerProxy(tokenManager_).getImplementationTypeAndTokenAddress();
+
+        // slither-disable-next-line controlled-delegatecall
+        (bool success, bytes memory data) = tokenHandler.delegatecall(
+            abi.encodeWithSelector(ITokenHandler.takeToken.selector, tmt, tokenAddress, tokenManager_, from, amount)
+        );
+        if (!success) revert TakeTokenFailed(data);
+        amount = abi.decode(data, (uint256));
+
+        ITokenManager(tokenManager_).addFlowOut(amount);
+
+        return (amount, tokenAddress);
+    }
+
+    function _giveToken(bytes32 tokenId, address to, uint256 amount) internal returns (uint256, address) {
+        address tokenManager_ = tokenManagerAddress(tokenId);
+        (uint256 tmt, address tokenAddress) = ITokenManagerProxy(tokenManager_).getImplementationTypeAndTokenAddress();
+
+        // slither-disable-next-line controlled-delegatecall
+        (bool success, bytes memory data) = tokenHandler.delegatecall(
+            abi.encodeWithSelector(ITokenHandler.giveToken.selector, tmt, tokenAddress, tokenManager_, to, amount)
+        );
+        if (!success) revert GiveTokenFailed(data);
+        amount = abi.decode(data, (uint256));
+
+        ITokenManager(tokenManager_).addFlowIn(amount);
```

Comment:
`addFlowIn` should be called before `giveToken`, since that's the amount ITS received in the payload regardless of what fraction was given to the user